pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+        """
+
+        Y = Enlist(Y)
+        if len(Y) != len(self._isets):
+            raise ValueError("Invalid length")
+        for iset, y in zip(self._isets, Y):
+            x_sub = x.getSubVector(iset)
+            if isinstance(y, Complex):
+                x_sub.set(y)
+            elif isinstance(y, OverloadedType):
+                y._ad_to_petsc(vec=x_sub)
+            else:
+                raise TypeError(f"Unexpected type: {type(y)}")
+            x_sub.restoreSubVector(iset, x_sub)
+
+
+class PETScOptions:
+    def __init__(self, options_prefix):
+        if PETSc is None:
+            raise RuntimeError("PETSc not available")
+
+        self._options_prefix = options_prefix
+        self._options = PETSc.Options()
+        self._keys = {}
+
+        def finalize_callback(options_prefix, options, keys):
+            for key in keys:
+                key = f"{options_prefix:s}{key}"
+                if key in options:
+                    del options[key]
+
+        finalize = weakref.finalize(
+            self, finalize_callback,
+            self._options_prefix, self._options, self._keys)
+        finalize.atexit = False
+
+    @property
+    def options_prefix(self):
+        return self._options_prefix
+
+    def __getitem__(self, key):
+        if key not in self._keys:
+            raise KeyError(key)
+        return self._options[f"{self.options_prefix:s}{key}"]
+
+    def __setitem__(self, key, value):
+        self._keys[key] = None
+        self._options[f"{self.options_prefix:s}{key:s}"] = value
+
+    def __delitem__(self, key):
+        del self._keys[key]
+        del self._options[f"{self.options_prefix:s}{key}"]
+
+    def clear(self):
+        for key in tuple(self._keys):
+            del self[key]
+
+    def update(self, other):
+        for key, value in other.items():
+            self[key] = value
+
+
+class TAOObjective:
+    """Utility class for computing functional values and associated
+    derivatives.
+
+    Args:
+        rf (ReducedFunctional): Defines the forward, and used to compute
+            derivative information.
+    """
+
+    def __init__(self, rf):
+        self._reduced_functional = rf
+
+    @property
+    def reduced_functional(self):
+        """:class:`.ReducedFunctional`. Defines the forward, and used to
+        compute derivative information.
+        """
+
+        return self._reduced_functional
+
+    def objective_gradient(self, M):
+        """Evaluate the forward, and compute a first derivative.
+
+        Args:
+            M (OverloadedType or Sequence[OverloadedType]): Defines the control
+                value.
+        Returns:
+            AdjFloat: The value of the functional.
+            OverloadedType or Sequence[OverloadedType]: The (dual space)
+                derivative.
+        """
+
+        M = Enlist(M)
+        J = self.reduced_functional(tuple(m._ad_copy() for m in M))
+        _ = self.reduced_functional.derivative()
```

Comment:
Can you please raise an issue on Pyadjoint pointing out that this hack is here.

Reply:
#164